File: src/main/Java/boj_step22/boj15652.java
import java.io.*;
import java.util.Stack;
import java.util.StringTokenizer;

public class boj15652 {
    private static int n = 0;
    private static int m = 0;
    private static Stack<Integer> result = new Stack<>();
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer tokenizer = new StringTokenizer(reader.readLine());
        n = Integer.parseInt(tokenizer.nextToken());
        m = Integer.parseInt(tokenizer.nextToken());
        StringBuilder builder = new StringBuilder();
        dfs(1, 0, builder);
        writer.append(builder).close();
    }
    private static void dfs(int tmp, int num, StringBuilder builder) {
        if (num == m) {
            result.forEach(x -> builder.append(x).append(' '));
            builder.append('\n');
            return;
        }
        for (int i = tmp; i <= n; i++) {
            result.push(i);
            dfs(i, num + 1, builder);
            result.pop();
        }
    }
}

File: src/main/Kotlin/boj_step22/boj15652.kt

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

private var n = 0
private var m = 0
private val result = Stack<Int>()
fun main() {
    val r = BufferedReader(InputStreamReader(System.`in`))
    val w = BufferedWriter(OutputStreamWriter(System.out))
    val t = StringTokenizer(r.readLine())
    n = t.nextToken().toInt()
    m = t.nextToken().toInt()
    val b = StringBuilder()
    dfs(1, 0, b)
    w.append(b).close()
}
private fun dfs(tmp: Int, num: Int, b: StringBuilder) {
    if (num == m) {
        result.forEach { b.append(it).append(' ') }
        b.appendLine()
        return
    }
    for (i in tmp .. n) {
        result.push(i)
        dfs(i, num + 1, b)
        result.pop()
    }
}